### PR TITLE
Two way irrigation coupling between ELM and MOSART

### PR DESCRIPTION
**Pull request by hydrotian**
Add two-way coupled irrigation scheme to land and river models. 
This function can be turn on/off by a flag in user_nl_clm.

ELM computes the surface water irrigation demand and sends to MOSART and
then MOSART determines the irrigation supply based on the demand and the
water availability and sends back to ELM to constrain the surface water irrigation.

[BFB]
[NML]

### CODE REVIEW

File: components/clm/src/biogeophys/CanopyHydrologyMod.F90 [hunk 1]
```diff
@@ -225,12 +242,92 @@ subroutine CanopyHydrology(bounds, &
...
+          qflx_surf_irrig      => veg_wf%qflx_surf_irrig_patch     , & ! Output: [real(r8) (:)   ]  actual surface water irrigation (mm/s)      
+          qflx_grnd_irrig      => veg_wf%qflx_grnd_irrig_patch     , & ! Output: [real(r8) (:)   ]  actual groundwater irrigation (mm/s)     
+          qflx_over_supply     => veg_wf%qflx_over_supply_patch      & ! Output: [real(r8) (:)   ]  the portion of supply that exceed total demand (mm/s)                 
           )
 
        ! Compute time step
        
        dtime = get_step_size()
+	   
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! temp solution for the irrigation mapping issue, if ELM and MOSART share the same grid, no such problem
```
Comment (bishtgautam) [inline components/clm/src/biogeophys/CanopyHydrologyMod.F90]: Please edit this line to only have a single `!`.

File: components/clm/src/biogeophys/CanopyHydrologyMod.F90 [hunk 2]
```diff
@@ -225,12 +242,92 @@ subroutine CanopyHydrology(bounds, &
...
+         irrig_rate (bounds%begp:bounds%endp), &  
+         irrig_rate_grid (bounds%begg:bounds%endg), &
+         p2c_scale_type='unity', c2l_scale_type= 'unity', l2g_scale_type='unity') 
+		 
+		! initialize the qflx_irrig_grid 
+         do gg = bounds%begg,bounds%endg
+            qflx_irrig_grid(gg) = 0
+         end do
+        
+        ! loop the pft and assign irrig_rate to qflx_irrig based on n_irrig_steps_left(p), which will be zero if current pft doesn't need irrigation or the current time is out of irrigation schedule
```
Comment (bishtgautam) [inline components/clm/src/biogeophys/CanopyHydrologyMod.F90]: Please split this extremely long comment into multiple lines

File: components/clm/src/biogeophys/CanopyHydrologyMod.F90 [hunk 3]
```diff
@@ -225,12 +242,92 @@ subroutine CanopyHydrology(bounds, &
...
+        if (total_sup_irrigrid .gt. 0._r8) then
+            !This is the ratio between supply over gridcells need water and supply over all gridcells
+            !This value should be ranging from 0 to 1 and will be used to concentrate the water supply into gridcells that need irrigation
+          adjust_f = total_sup_irrigrid/total_sup 
+        end if	
+
+        if (adjust_f < 1.0e-3_r8) then
+            adjust_f = 1 ! if irrigated area is too small, don't adjust because it would generate huge irrigation rate in a very small area
+        end if            
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
```
Comment (bishtgautam) [inline components/clm/src/biogeophys/CanopyHydrologyMod.F90]: Please delete this line

File: components/clm/src/biogeophys/CanopyHydrologyMod.F90 [hunk 4]
```diff
@@ -225,12 +242,92 @@ subroutine CanopyHydrology(bounds, &
...
+		 		   
+        !initialize the total supply for all grids over the processer and total supply for grids that need irrigation 
+        total_sup_irrigrid = 0._r8
+        total_sup = 0._r8
+        
+        currentg = bounds%begg
+        do gg = 1, gridnum
+          g = currentg + gg - 1
+		  
+		  !if (atm2lnd_vars%supply_grc(g)>0) then
```
Comment (bishtgautam) [inline components/clm/src/biogeophys/CanopyHydrologyMod.F90]: Please delete this commented out line

File: components/clm/src/biogeophys/CanopyHydrologyMod.F90 [hunk 5]
```diff
@@ -331,14 +430,75 @@ subroutine CanopyHydrology(bounds, &
...
+ 
+               if (qflx_irrig(p) > 0._r8) then	!this pft needs water	   
+               qflx_surf_irrig(p) = atm2lnd_vars%supply_grc(g)/adjust_f/pgwgt(p)    ! original supply at grid level (mm/s) concentrate
+                                                                                    ! to grid cells that need water  and then project 
+																		            ! to pft level     
+               if (pgwgt(p) < 1.0e-16_r8) then ! avoid super small fraction of irrigated land 
+                   qflx_surf_irrig(p) = atm2lnd_vars%supply_grc(g)/adjust_f
+               end if      
+               
+               !!!!!!!!!!!!!! use supply_frac to calculate surf_irrg
```
Comment (bishtgautam) [inline components/clm/src/biogeophys/CanopyHydrologyMod.F90]: Please delete this commented out lines

File: components/clm/src/biogeophys/CanopyHydrologyMod.F90 [hunk 6]
```diff
@@ -331,14 +430,75 @@ subroutine CanopyHydrology(bounds, &
...
+                   qflx_prec_grnd_rain(p) = qflx_prec_grnd_rain(p) + qflx_real_irrig(p) + qflx_over_supply(p) !applying irrigation, the over supply is included to balance water
+               
+               
+             else !this pft doesn't need water             
+               qflx_prec_grnd_rain(p) = qflx_prec_grnd_rain(p)
+               qflx_real_irrig(p) = 0! this should be zero, just leave it here for testing added by Tian 2/27/2018
+               qflx_surf_irrig(p) = 0
+               qflx_grnd_irrig(p) = 0
+               qflx_over_supply(p) = 0
+               !if (qflx_irrig(p) > 0) then
```
Comment (bishtgautam) [inline components/clm/src/biogeophys/CanopyHydrologyMod.F90]: Please delete this commented out lines

File: components/clm/src/biogeophys/HydrologyDrainageMod.F90 [hunk 7]
```diff
@@ -148,6 +152,9 @@ subroutine HydrologyDrainage(bounds,    &
         call ep_betr%DiagDrainWaterFlux(num_hydrologyc, filter_hydrologyc)
         call ep_betr%RetrieveBiogeoFlux(bounds, 1, nlevsoi, waterflux_vars=waterflux_vars)
       endif
+
```
Comment (bishtgautam) [inline components/clm/src/biogeophys/HydrologyDrainageMod.F90]: Please delete this commented out lines

File: components/clm/src/biogeophys/HydrologyDrainageMod.F90 [hunk 8]
```diff
@@ -275,7 +282,8 @@ subroutine HydrologyDrainage(bounds,    &
          qflx_runoff(c) = qflx_drain(c) + qflx_surf(c)  + qflx_h2osfc_surf(c) + qflx_qrgwl(c) + qflx_drain_perched(c)
 
          if ((lun_pp%itype(l)==istsoil .or. lun_pp%itype(l)==istcrop) .and. col_pp%active(c)) then
-            qflx_runoff(c) = qflx_runoff(c) - qflx_irrig(c)
+            !qflx_runoff(c) = qflx_runoff(c) - qflx_irrig(c)
```
Comment (bishtgautam) [inline components/clm/src/biogeophys/HydrologyDrainageMod.F90]: Please delete this commented out line

File: components/clm/src/biogeophys/HydrologyDrainageMod.F90 [hunk 9]
```diff
@@ -285,6 +293,8 @@ subroutine HydrologyDrainage(bounds,    &
 
       end do
 
+      !call WaterTable(bounds, num_hydrologyc, filter_hydrologyc, num_urbanc, filter_urbanc, &
```
Comment (bishtgautam) [inline components/clm/src/biogeophys/HydrologyDrainageMod.F90]: Please delete this commented out lines

File: components/clm/src/biogeophys/SoilHydrologyMod.F90 [hunk 10]
```diff
@@ -689,6 +697,24 @@ subroutine WaterTable(bounds, num_hydrologyc, filter_hydrologyc, num_urbanc, fil
           rous=max(rous,0.02_r8)
 
           !--  water table is below the soil column  --------------------------------------
+		  g = col_pp%gridcell(c)
+          l = col_pp%landunit(c)
+
+       !   local runoff used for irrigation , commented out because it's been taken care of in MOSART-WM 
+       !   qcharge_temp = qcharge(c)
```
Comment (bishtgautam) [inline components/clm/src/biogeophys/SoilHydrologyMod.F90]: Please delete this commented out lines

File: components/clm/src/data_types/VegetationDataType.F90 [hunk 11]
```diff
@@ -5348,17 +5355,55 @@ subroutine veg_wf_init(this, begp, endp)
...
     allocate(this%n_irrig_steps_left     (begp:endp))             ; this%n_irrig_steps_left   (:)   = 0
     
     !-----------------------------------------------------------------------
     ! initialize history fields for select members of veg_wf
     !-----------------------------------------------------------------------
-    this%qflx_irrig(begp:endp) = spval
-    call hist_addfld1d (fname='QIRRIG', units='mm/s', &
-         avgflag='A', long_name='water added through irrigation', &
-         ptr_patch=this%qflx_irrig)
+    !this%qflx_irrig(begp:endp) = spval
```
Comment (bishtgautam) [inline components/clm/src/data_types/VegetationDataType.F90]: Please delete this commented out lines

File: components/mosart/bld/namelist_files/namelist_defaults_mosart.xml [hunk 12]
```diff
@@ -32,4 +39,34 @@ for the CLM data in the CESM distribution
 <frivinp_rtm rof_grid="r05" >rof/mosart/MOSART_global_half_20180721a.nc</frivinp_rtm>
 <frivinp_rtm rof_grid="NLDAS" >rof/mosart/MOSART_NLDAS_8th.nc</frivinp_rtm>
 
+<!-- WRM -->
+
+<paraFile rof_grid="NLDAS" >rof/mosart/US_reservoir_8th_NLDAS3_c20161220_updated_20170314.nc</paraFile>
```
Comment (bishtgautam) [inline components/mosart/bld/namelist_files/namelist_defaults_mosart.xml]: Has this file been added to e3sm inputdata sever?

File: components/mosart/src/riverroute/RtmRestFile.F90 [hunk 13]
```diff
@@ -23,6 +23,10 @@ module RtmRestFile
   use RunoffMod     , only : rtmCTL
   use RtmIO       
   use RtmDateTime
+!#ifdef INCLUDE_WRM
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/RtmRestFile.F90]: Please delete this commented out lines

File: components/mosart/src/inundation/MOSARTinund_Core_MOD.F90 [hunk 14]
```diff
@@ -0,0 +1,750 @@
+
+MODULE MOSARTinund_Core_MOD
+
+!#ifdef INCLUDE_INUND
```
Comment (bishtgautam) [inline components/mosart/src/inundation/MOSARTinund_Core_MOD.F90]: Please delete the `!#ifdef`

**Comment by rljacob** [comment]
The description above should also describe the code changes/additions.

File: components/mosart/bld/namelist_files/namelist_defaults_mosart.xml (hunk 12 above)
Comment (hydrotian) [reply components/mosart/bld/namelist_files/namelist_defaults_mosart.xml]: I don't think I have the permission to add this file to the input data server. Could you help me uploading it? You can find it in Constance input data directory.

**Comment by wlin7** [comment]
@bishtgautam , looks like @hydrotian has addressed all the suggested changes. Is this PR ready to go in now?

**Comment by bishtgautam** [comment]
@wlin7 This PR is ready to go and I will merge it to next today.

**Comment by dqwu** [comment]
This PR (merged to next) has caused the following test to fail on anlworkstation:
SMS_D_Ln5.ne4_ne4.FC5AV1C-L

```
Error termination. Backtrace:
At line 382 of file components/clm/src/biogeophys/WaterfluxType.F90
Fortran runtime error: Index '6182' of dimension 1 of array 'this%qflx_irr_demand_col' above upper bound of 6181
```
The code block including line 382 is:
```
    do p = bounds%begp, bounds%endp
       l = veg_pp%landunit(p)

       if (lun_pp%itype(l)==istsoil) then
          this%n_irrig_steps_left_patch(p) = 0
          this%irrig_rate_patch(p)         = 0.0_r8
          this%qflx_irr_demand_col(c)  = 0._r8
       end if
    end do
```
Should line 382 use p as the index (instead of c)?
`this%qflx_irr_demand_col(p)  = 0._r8`

**Comment by dqwu** [comment]
PS, I think this feature branch (tizhou/two-way-irrigation) should be rebased with latest master
1) It is 578 commits behind master.
2) There are merging conflicts to be resolved:
components/clm/src/biogeophys/CanopyHydrologyMod.F90
3) There are 16 commits in this PR, but 8 of them are duplicate commits which have already been merged to master by PR #3012:
https://github.com/E3SM-Project/E3SM/pull/3074/commits
https://github.com/E3SM-Project/E3SM/pull/3012/commits

**Comment by bishtgautam** [comment]
@dqwu Thanks for your catching the error in  `qflx_irr_demand_col(c)` and also pointing out that bunch of commits in this branch are already on master. I will work with the developer and rebase the branch of the latest master. I will also revert this PR on next so anlworkstation reports a pass.

**Comment by bishtgautam** [comment]
@dqwu I don't have access to anlworkstation. What compiler is being used on anlworkstation? In the meantime, I will run the SMS_D_Ln5.ne4_ne4.FC5AV1C-L test on Cori.

**Comment by bishtgautam** [comment]
@dqwu I can reproduce the failure of the test on Cori.

**Comment by dqwu** [comment]
@bishtgautam 
I used GNU (version 8.2) on anlworkstation.
The default compiler on Cori is Intel but GNU can also be used (--compiler=gnu)
It looks that this failure can be reproduced with both GNU and Intel.

**Comment by dqwu** [comment]
@bishtgautam 
That test still failed on anlworkstation with the new commit "Fix the checking if value is NaN":
```
[0] Program received signal SIGFPE: Floating-point exception - erroneous arithmetic operation.
[0]
[0] Backtrace for this error:
[0] #0  0x7fc34e241caf in ???
[0] #1  0x1fc0991 in __shr_infnan_mod_MOD_shr_infnan_isnan_double
[0]     at bld/gnu/mpich/debug/nothreads/mct/mct/noesmf/c1a1l1i1o1r1g1w1i1e1/csm_share/shr_infnan_mod.F90:227
[0] #2  0x16655cb in __canopyhydrologymod_MOD_canopyhydrology
[0]     at components/clm/src/biogeophys/CanopyHydrologyMod.F90:265
[0] #3  0x1286384 in __clm_driver_MOD_clm_drv
[0]     at components/clm/src/main/clm_driver.F90:1296
[0] #4  0x126fdf4 in __lnd_comp_mct_MOD_lnd_run_mct
[0]     at components/clm/src/cpl/lnd_comp_mct.F90:509
[0] #5  0x42819a in __component_mod_MOD_component_run
[0]     at cime/src/drivers/mct/main/component_mod.F90:714
```

**Comment by bishtgautam** [comment]
@dqwu, @hydrotian has fixed the failing SMS_D_Ln5.ne4_ne4.FC5AV1C-L test in the latest commit. I have verified that that test passes on Cori. Do you want to double check if that test passes on anlworkstation? Or, should I go ahead and merge this PR to next?

**Comment by rljacob** [comment]
@dqwu note that anlworkstation is not an officially supported platform for E3SM.  If you find a bug there, make sure it can be produced on an official platform before reporting.

**Comment by bishtgautam** [comment]
@hydrotian Can you address @rljacob's [comment](https://github.com/E3SM-Project/E3SM/pull/3074#pullrequestreview-288855236)? Thanks.

**Comment by dqwu** [comment]
> @dqwu, @hydrotian has fixed the failing SMS_D_Ln5.ne4_ne4.FC5AV1C-L test in the latest commit. I have verified that that test passes on Cori. Do you want to double check if that test passes on anlworkstation? Or, should I go ahead and merge this PR to next?

I have verified that it passes on anlworkstation with the latest commit.

**Comment by hydrotian** [comment]
I have updated the descriptions of this PR in Confluence page: https://acme-climate.atlassian.net/wiki/spaces/ED/pages/1000276179/W5+Two-way+Coupled+Irrigation+Design+Document

**Comment by rljacob** [comment]
Thats great but I specifically want the description on this PR to have more detail.  You don't need to go file-by-file but say in general what was added/changed.  The PR description at the top will go in the git log which is more permanent then Confluence.

**Comment by hydrotian** [comment]
Description updated. Also made a small commit to change 0._r8 to nan for some new variable allocations.

**Comment by bishtgautam** [comment]
@rljacob, Are you okay with the updated PR description?

**Comment by rljacob** [comment]
I'd like a test for this feature to be added to the integration suite.

**Pull request merged**